karapace/schema_reader.py

```diff
@@ -234,11 +231,8 @@ def _get_beginning_offset(self) -> int:
         assert self.consumer is not None, "Thread must be started"
 
         try:
-            offsets = self.consumer.beginning_offsets([TopicPartition(self.config["topic_name"], 0)])
-            # Offset in the response is the offset for last offset.
-            # Reduce by one for matching on startup.
-            beginning_offset = list(offsets.values())[0] - 1
-            return beginning_offset
+            beginning_offset, _ = self.consumer.get_watermark_offsets(TopicPartition(self.config["topic_name"], 0))
+            return beginning_offset - 1
```

Comment:
Should this be reduced by 1? 🤔

Reply:
weird, I cannot understand why trivially. If you remove the `-1` and run again the tests some of them are failing?

Reply:
also we need to check that for an empty topic returns 0 and from a topic with 1 message return always 0

Reply:
Having or not having the `-1` there makes no difference in the tests, interestingly.
I added some tests for the consumer for the purpose of documenting this behaviour - `(0, 0)` is returned on an empty topic by `get_watermark_offsets`, so I think the `-1`s are correct here

Reply:
I would say that we should be able to distinguish between an empty topic and a topic with 1 message. IMO should give back 2 different reply

Reply:
As discussed separately with @aiven-anton and @jjaakola-aiven, this offset decrement is a result of the schema reader and offset watcher evolution and has no real effect towards behaviour (unlike decrementing the end offset). See also `OFFSET_UNINITIALIZED` and `OFFSET_EMPTY` towards the top of this file. This and the two constants can be eventually cleaned up and simplified.

Related explanation in the backups code: https://github.com/Aiven-Open/karapace/pull/781

Reply:
question, not requirement. Do you think would be useful to paste the explaination also here? I feel that its something that could save time to someone wondering why in the future

Reply:
@eliax1996 I added a comment here